laplace/curvature/curvlinops.py

```diff
@@ -0,0 +1,128 @@
+import torch
+import numpy as np
+
+import curvlinops as cvls
+
+from laplace.curvature import CurvatureInterface, GGNInterface, EFInterface
+from laplace.utils import Kron
+
+from collections import UserDict
+
+from typing import *
+
+
+class CurvlinopsInterface(CurvatureInterface):
+    """Interface for Curvlinops backend. <https://github.com/f-dangel/curvlinops>
+    """
+    def __init__(self, model, likelihood, last_layer=False, subnetwork_indices=None):
+        super().__init__(model, likelihood, last_layer, subnetwork_indices)
+
+    @property
+    def _kron_fisher_type(self):
+        raise NotImplementedError
+
+    @property
+    def _linop_context(self):
+        raise NotImplementedError
+
+    @staticmethod
+    def _rescale_kron_factors(kron, M, N):
+        # Renormalize Kronecker factor to sum up correctly over N data points with batches of M
+        # for M=N (full-batch) just M/N=1
+        for F in kron.kfacs:
+            if len(F) == 2:
+                F[1] *= M/N
+        return kron
+
+    def _get_kron_factors(self, linop):
+        kfacs = list()
+        for name, module in self._model.named_modules():
+            if name not in linop._mapping.keys():
+                continue
+
+            A = linop._input_covariances[name]
+            B = linop._gradient_covariances[name]
+
+            if hasattr(module, 'bias') and module.bias is not None:
+                kfacs.append([B, A])
+                kfacs.append([B])
+            elif hasattr(module, 'weight'):
+                p, q = np.prod(B.shape), np.prod(A.shape)
+                if p == q == 1:
+                    kfacs.append([B * A])
+                else:
+                    kfacs.append([B, A])
+            else:
+                raise ValueError(f'Whats happening with {module}?')
+        return Kron(kfacs)
+
+    def kron(self, X, y, N, **kwargs):
+        linop = cvls.KFACLinearOperator(
+            self.model, self.lossfunc, self.params, [(X, y)],
+            fisher_type=self._kron_fisher_type,
+            loss_average=None,  # Since self.lossfunc is sum
+            separate_weight_and_bias=True,
+        )
+        linop._compute_kfac()
+
+        kron = self._get_kron_factors(linop)
+        kron = self._rescale_kron_factors(kron, len(y), N)
+        kron *= self.factor
+
+        # Correction to match BackPACK & ASDL results
+        for factors in kron.kfacs:
+            for i in range(len(factors)):
+                factors[i] /= 2
+
+        loss = self.lossfunc(self.model(X), y)
+
+        return self.factor * loss.detach(), kron
```

Comment:
@runame I found the culprit. Curvlinops' KFAC factors differ are twice as large as BackPACK's and ASDL's. So I put a correction in L72-75 above. 

I'm not sure what's happening. Please propose a more elegant fix.

Reply:
Fixed in [this commit](https://github.com/aleximmer/Laplace/pull/146/commits/8d8b8bb4eb90df05efcf79f6b8da8d67ecb29127). The issue was that `linop._compute_kfac()` was called twice, so that all Kronecker factors were added to itself once, resulting in the factor of 2. This can be avoided (also more efficient) by setting `check_deterministic=False`.